crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -332,6 +330,157 @@ impl Drop for BlobVec {
     }
 }
 
+pub(super) struct BlobBox {
+    item_layout: Layout,
+    is_present: bool,
+    // the `data` ptr's layout is always `array_layout(item_layout, capacity)`
+    data: NonNull<u8>,
+    // None if the underlying type doesn't need to be dropped
+    drop: Option<unsafe fn(OwningPtr<'_>)>,
+}
+
+// We want to ignore the `drop` field in our `Debug` impl
+impl std::fmt::Debug for BlobBox {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BlobBox")
+            .field("item_layout", &self.item_layout)
+            .field("is_present", &self.is_present)
+            .field("data", &self.data)
+            .finish()
+    }
+}
+
+impl BlobBox {
+    /// # Safety
+    ///
+    /// `drop` should be safe to call with an [`OwningPtr`] pointing to any item that's been pushed into this [`BlobVec`].
+    ///
+    /// If `drop` is `None`, the items will be leaked. This should generally be set as None based on [`needs_drop`].
+    ///
+    /// [`needs_drop`]: core::mem::needs_drop
+    pub unsafe fn new(item_layout: Layout, drop: Option<unsafe fn(OwningPtr<'_>)>) -> BlobBox {
+        if item_layout.size() == 0 {
+            let align = NonZeroUsize::new(item_layout.align()).expect("alignment must be > 0");
+            BlobBox {
+                data: bevy_ptr::dangling_with_align(align),
+                is_present: false,
+                item_layout,
+                drop,
+            }
+        } else {
+            let data = std::alloc::alloc(item_layout);
+            BlobBox {
+                data: NonNull::new(data).unwrap_or_else(|| handle_alloc_error(item_layout)),
+                is_present: false,
+                item_layout,
+                drop,
+            }
+        }
+    }
+
+    #[inline]
+    pub fn is_present(&self) -> bool {
+        self.is_present
+    }
+
+    /// # Safety
+    /// - index must be in bounds
+    /// - the memory in the [`BlobVec`] starting at index `index`, of a size matching this [`BlobVec`]'s
+    /// `item_layout`, must have been previously allocated.
+    #[inline]
+    pub unsafe fn initialize(&mut self, value: OwningPtr<'_>) {
+        debug_assert!(!self.is_present());
+        std::ptr::copy_nonoverlapping::<u8>(
+            value.as_ptr(),
+            self.data.as_ptr(),
+            self.item_layout.size(),
+        );
+        self.is_present = true;
+    }
+
+    /// # Safety
+    /// - index must be in-bounds
+    /// - the memory in the [`BlobVec`] starting at index `index`, of a size matching this
+    /// [`BlobVec`]'s `item_layout`, must have been previously initialized with an item matching
+    /// this [`BlobVec`]'s `item_layout`
+    /// - the memory at `*value` must also be previously initialized with an item matching this
+    /// [`BlobVec`]'s `item_layout`
+    pub unsafe fn replace(&mut self, value: OwningPtr<'_>) {
+        let ptr = self.data.as_ptr();
+        self.is_present = false;
```

Comment:
If we stored the value of `self.is_present` in a temporary first, we could make this function safe for when the BlobBox is not yet initialized. Would that be a useful property? It would make code like `insert_with_ticks` a little bit simpler, since we could always call `replace` instead.

Reply:
We could do so, and it would move the unsafety into the function, but changing this without changing ResourceData will add an extra branch where there doesn't need to be one.